### PR TITLE
Accept step object in dependson list

### PR DESCRIPTION
**Pull request by jerrypeng7773**
*Issue #, if available:*

*Description of changes:*

support step object in step depends on list

*Testing done:*

unit tested

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by staubhp — DISMISSED** [review]

File: src/sagemaker/workflow/steps.py [hunk 1]
```diff
@@ -63,12 +63,13 @@ class Step(Entity):
...
         step_type (StepTypeEnum): The type of the step.
-        depends_on (List[str]): The list of step names the current step depends on
+        depends_on (List[str] or List[Step]): The list of step names or step
+            instances the current step depends on
     """
 
     name: str = attr.ib(factory=str)
     step_type: StepTypeEnum = attr.ib(factory=StepTypeEnum.factory)
-    depends_on: List[str] = attr.ib(default=None)
+    depends_on: Union[List[str], List["Step"]] = attr.ib(default=None)
```
Comment (navaj0) [inline src/sagemaker/workflow/steps.py:72]: List["Step"] -> List[Step]

File: src/sagemaker/workflow/steps.py [hunk 2]
```diff
@@ -105,6 +108,19 @@ def ref(self) -> Dict[str, str]:
         """Gets a reference dict for steps"""
         return {"Name": self.name}
 
+    @staticmethod
+    def _resolve_depends_on(depends_on_list: Union[List[str], List["Step"]]):
```
Comment (navaj0) [inline src/sagemaker/workflow/steps.py:112]: List["Step"] -> List[Step]

Comment (jerrypeng7773) [reply src/sagemaker/workflow/steps.py:112]: in py37 in order to infer type to self class, we needs to do `from __future__ import annotations`, and this is not supported in py36. The pythonSDK supports both 37 and 36, so I think we need to stick with string "Step" for a while.

https://stackoverflow.com/questions/33533148/how-do-i-type-hint-a-method-with-the-type-of-the-enclosing-class

**Review by EthanShouhanCheng — DISMISSED** [review]
LGTM

**Review by navaj0 — APPROVED** [review]

**Review by EthanShouhanCheng — APPROVED** [review]

**Review by staubhp — APPROVED** [review]

**Review by shreyapandit — DISMISSED** [review]

**Review by ahsan-z-khan — APPROVED** [review]

**Review by shreyapandit — APPROVED** [review]

**Pull request merged**